vispy/visuals/_scalable_textures.py

```diff
@@ -0,0 +1,441 @@
+
+    This class includes optimizations where image data is not re-normalized
+    if the previous normalization can still be used to visualize the data
+    with the new color limits.
+
+    This class should only be used internally. For similar features where
+    scaling occurs on the GPU see
+    :class:`vispy.visuals._scalable_textures.GPUScaledTextureMixin`.
+
+    To use this mixin, a subclass should be created to combine this mixin with
+    the texture class being used. Existing subclasses already exist in this
+    module. Note that this class **must** appear first in the subclass's parent
+    classes so that its ``__init__`` method is called instead of the parent
+    Texture class.
+
+    """
+
+    def __init__(self, data=None, **texture_kwargs):
+        self._data_limits = None
+        # Call the __init__ of the mixin base class
+        super().__init__(data, **texture_kwargs)
+
+    def _clim_outside_data_limits(self, cmin, cmax):
+        if self._data_limits is None:
+            return False
+        return cmin < self._data_limits[0] or cmax > self._data_limits[1]
+
+    def set_clim(self, clim):
+        """Set clim and return if a texture update is needed."""
+        need_texture_upload = False
+        # NOTE: Color limits are not checked against data type limits
+        if isinstance(clim, str):
+            if clim != 'auto':
+                raise ValueError('clim must be "auto" if a string')
+            need_texture_upload = True
+            self._clim = clim
+        else:
+            try:
+                cmin, cmax = clim
+            except (ValueError, TypeError):
+                raise ValueError('clim must have two elements')
+            if self._clim_outside_data_limits(cmin, cmax):
+                need_texture_upload = True
+            self._clim = (cmin, cmax)
+        return need_texture_upload
+
+    @property
+    def clim_normalized(self):
+        """Normalize current clims to match texture data inside the shader.
+
+        If data is scaled on the CPU then the texture data will be in the range
+        0-1 in the _build_texture() method. Inside the fragment shader the
+        final contrast adjustment will be applied based on this normalized
+        ``clim``.
+
+        """
+        range_min, range_max = self._data_limits
+        clim_min, clim_max = self.clim
+        clim_min = (clim_min - range_min) / (range_max - range_min)
+        clim_max = (clim_max - range_min) / (range_max - range_min)
+        return clim_min, clim_max
+
+    @staticmethod
+    def _scale_data_on_cpu(data, clim, copy=True):
+        if copy:
+            should_cast_to_f32(data.dtype)
+            data = np.array(data, dtype=np.float32, copy=copy)
+        elif not copy and not np.issubdtype(data.dtype, np.floating):
+            raise ValueError("Data must be of floating type for no copying to occur.")
+
+        if clim[0] == clim[1]:
+            if clim[0] != 0:
+                data /= clim[0]
```

Comment:
Adding comment and turn the numpy-division into a (faster) numpy-multiplication:
```suggestion
        if clim[0] == clim[1]:
            # Prevent zero division when all elements have the same value
            if clim[0] != 0:
                data *= 1 / clim[0]
```